src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -824,6 +899,18 @@ private static uint HandleEventShutdownComplete(State state, ref StreamEvent evt
                 state.ShutdownCompletionSource.SetResult();
             }
 
+            // Dispose was called before complete event.
+            bool releaseHandles = Interlocked.Exchange(ref state.ShutdownDone, 2) == 1;
```

Comment:
Isn't it the case that we always get this ShutdownComplete event? So then can't we always release the handles here?

Reply:
I was thinking about it. However in some cases when can ShutdownComplete without us asking for it .e.g. Disposing. With that, attempt to shutdown/reset will throw. So we would need careful synchronization before (and perhaps during) every operation.

Reply:
It seems like the handle is only needed for event callbacks, right? I don't think other code should be accessing it. So I don't think there's any real synchronization needed here, is there?

Reply:
I'm not sure I understand the comment. The cleanup does more than just handle and we also pass the handle in functions like Shutdown().